ebmbot/dispatcher.py

```diff
@@ -0,0 +1,182 @@
+import os
+import shlex
+import subprocess
+import time
+from datetime import datetime, timezone
+from multiprocessing import Process
+
+import requests
+from slackbot.slackclient import SlackClient
+
+from . import job_configs, scheduler, settings
+from .logger import logger
+
+
+def run():  # pragma: no cover
+    """Start the dispatcher running."""
+
+    slack_client = SlackClient(settings.SLACKBOT_API_TOKEN)
+
+    while True:
+        run_once(slack_client, job_configs.config)
+        time.sleep(1)
+
+
+def run_once(slack_client, config):
+    """Clear any expired suppressions, then start a new subprocess for each
+    available job.
+
+    We collect and return started processes so that we can wait for them to
+    finish in tests before asserting the tests have done anything.
+    """
+
+    scheduler.remove_expired_suppressions()
+
+    processes = []
+
+    while True:
+        job_id = scheduler.reserve_job()
+        if job_id is None:
+            break
+        job_dispatcher = JobDispatcher(slack_client, job_id, config)
+        processes.append(job_dispatcher.start_job())
+
+    return processes
+
+
+class JobDispatcher:
+    def __init__(self, slack_client, job_id, config):
+        logger.info("starting job", job_id=job_id)
+        self.slack_client = slack_client
+        self.job = scheduler.get_job(job_id)
+
+        namespace = self.job["type"].split("_")[0]
+        self.cwd = os.path.join(settings.WORKSPACE_DIR, namespace)
+        self.fabfile_url = config["fabfiles"].get(namespace)
+
+        job_config = config["jobs"][self.job["type"]]
+        self.run_args = build_run_args(
+            job_config["run_args_template"], self.job["args"]
+        )
+        self.report_stdout = bool(job_config.get("report_stdout"))
+
+    def start_job(self):
+        """Start running the job in a new subprocess."""
+
+        p = Process(target=self.do_job)
+        p.start()
+        return p
+
+    def do_job(self):
+        """Run the job."""
+
+        self.set_up_cwd()
+        self.set_up_log_dir()
+        self.notify_start()
+        rc = self.run_command()
+        scheduler.mark_job_done(self.job["id"])
+        self.notifiy_end(rc)
+
+    def run_command(self):
+        """Run the bash command, writing stdout/stderr to separate files."""
+
+        logger.info("run_command {")
+        logger.info(
+            "run_command",
+            job=self.job,
+            run_args=self.run_args,
+            cwd=self.cwd,
+            stdout_path=self.stdout_path,
+            stderr_path=self.stdout_path,
+        )
+
+        with open(self.stdout_path, "w") as stdout, open(
+            self.stderr_path, "w"
+        ) as stderr:
+            try:
+                rv = subprocess.run(
+                    self.run_args, cwd=self.cwd, stdout=stdout, stderr=stderr
```

Comment:
If it were me I would probably just have done `stderr=subprocess.STDOUT` and captured everything to one file. But that's mainly because I'm lazy and you've gone to trouble of doing things correctly.